controllers/azurejson_machinepool_controller.go

```diff
@@ -0,0 +1,156 @@
+	ctx, cancel := context.WithTimeout(context.Background(), reconciler.DefaultedLoopTimeout(r.ReconcileTimeout))
+	defer cancel()
+	log := r.Log.WithValues("namespace", req.Namespace, "AzureJSONMachinePool", req.Name)
+
+	// Fetch the AzureMachine instance
+	azureMachinePool := expv1.AzureMachinePool{}
+	err := r.Get(ctx, req.NamespacedName, &azureMachinePool)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("object was not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Fetch the CAPI MachinePool.
+	machinePool, err := GetOwnerMachinePool(ctx, r.Client, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if machinePool == nil {
+		log.Info("MachinePool Controller has not yet set OwnerRef")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("machinePool", machinePool.Name)
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machinePool.ObjectMeta)
+	if err != nil {
+		log.Info("MachinePool is missing cluster label or cluster does not exist")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	_, kind := infrav1.GroupVersion.WithKind("AzureCluster").ToAPIVersionAndKind()
+
+	// only look at azure clusters
+	if cluster.Spec.InfrastructureRef.Kind != kind {
+		log.WithValues("kind", cluster.Spec.InfrastructureRef.Kind).Info("infra ref was not an AzureCluster")
+		return ctrl.Result{}, nil
+	}
+
+	// fetch the corresponding azure cluster
+	azureCluster := &infrav1.AzureCluster{}
+	azureClusterName := types.NamespacedName{
+		Namespace: req.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+
+	err = r.Get(ctx, azureClusterName, azureCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("object was not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Create the scope.
+	clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+		Client:       r.Client,
+		Logger:       log,
+		Cluster:      cluster,
+		AzureCluster: azureCluster,
+	})
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+
+	apiVersion, kind := infrav1.GroupVersion.WithKind("AzureMachinePool").ToAPIVersionAndKind()
+	owner := metav1.OwnerReference{
+		APIVersion: apiVersion,
+		Kind:       kind,
+		Name:       azureMachinePool.GetName(),
+		UID:        azureMachinePool.GetUID(),
+	}
+
+	newSecret, err := GetCloudProviderSecret(
+		clusterScope,
+		azureMachinePool.Namespace,
+		azureMachinePool.Name,
+		infrav1.VMIdentityNone,
+		"",
+	)
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create cloud provider config")
+	}
+
+	if err := reconcileAzureSecret(ctx, log, r.Client, owner, newSecret); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to reconcile azure secret")
```

Comment:
Could we maybe add an event on the corresponding machinepool/machine for the secret failure?

Reply:
added, sorry I accidentally rebased before letting you review the newly added bits